packages/ui/src/providers/substrate/provider.tsx

```diff
@@ -0,0 +1,77 @@
+import React, { Dispatch, ReactNode, useReducer } from 'react'
+import { ApiPromise, WsProvider } from '@polkadot/api'
+import { web3Accounts, web3Enable } from '@polkadot/extension-dapp'
+import keyring from '@polkadot/ui-keyring'
+import jsonrpc from '@polkadot/types/interfaces/jsonrpc'
+import { Action, reducer, State, types } from './reducer'
+import { SubstrateContext } from './context'
+
+const socket = 'ws://127.0.0.1:9944'
+
+const connect = (state: State, dispatch: Dispatch<Action>) => {
+  const { apiState, socket, jsonrpc, types } = state
+  // We only want this function to be performed once
+  if (apiState) return
+
+  dispatch({ type: 'CONNECT_INIT' })
+
+  const provider = new WsProvider(socket)
+  const apiPromise = new ApiPromise({ provider, types, rpc: jsonrpc })
+
+  // Set listeners for disconnection and reconnection event.
+  apiPromise.on('connected', () => {
+    dispatch({ type: 'CONNECT', payload: apiPromise })
+    // `ready` event is not emitted upon reconnection and is checked explicitly here.
+    apiPromise.isReady.then(() => dispatch({ type: 'CONNECT_SUCCESS' }))
+  })
+  apiPromise.on('ready', () => dispatch({ type: 'CONNECT_SUCCESS' }))
+  apiPromise.on('error', (err) => dispatch({ type: 'CONNECT_ERROR', payload: err }))
+}
+
+let loadAccts = false
+
+const loadAccounts = (state: State, dispatch: Dispatch<Action>) => {
+  const asyncLoadAccounts = async () => {
+    dispatch({ type: 'LOAD_KEYRING' })
+    try {
+      await web3Enable('Pioneer')
+      let allAccounts = await web3Accounts()
```

Comment:
For loading accounts we may want to consider some sort of subscription, as in [apps](https://github.com/polkadot-js/apps/blob/master/packages/react-hooks/src/useAccounts.ts#L21). That way when account is added in the extension it is reflected on the page without the need for reloading. Can come as next PR though.

Reply:
I see if this will be required for #10 as well.